File: yiya-crawler-deploy/src/utils/fileHandler.js
// import fs from 'fs/promises';
import path from 'path';
import { extractDomain } from '../services/storage.js';
import { createYiyaJob } from '../services/yiya.js';
import { uploadToOSS } from '../utils/ossClient.js';
import { WebPage } from '../models/WebPage.js';
import { logger } from '../utils/logger.js';
import { SPIDER_SITE_TYPE } from '../config/constants.js';
import fs from 'fs';
import { pipeline } from 'stream';
import { promisify } from 'util';
import got from 'got';

// 临时PDF存储目录
const TEMP_DIR = 'downloads';
const ATTACHMENT_DIR = 'downloads/attachments';
const streamPipeline = promisify(pipeline);

/**
 * 确保目录存在
 * @param {string} dirPath 目录路径
 */
export const ensureDir = async (dirPath) => {
  try {
    await fs.access(dirPath);
  } catch {
    await fs.mkdir(dirPath, { recursive: true });
  }
};

/**
 * 生成唯一的文件名
 * @param {string} url 文件URL
 * @param {string} ext 文件扩展名
 */
export const generateUniqueName = (url, ext = 'pdf') => {
  const timestamp = Date.now();
  const urlHash = Buffer.from(url).toString('base64').slice(0, 10);
  return `${timestamp}_${urlHash}.${ext}`;
};


/**
 * 生成唯一文件名
 * @param {string} url 网页URL
 */
export const generatePdfName = (url) => {
  const timestamp = Date.now();
  const hostname = new URL(url).hostname;
  return `${hostname}_${timestamp}.pdf`;
};

/**
 * 清理临时文件
 */
export const cleanupTempFiles = async () => {
  try {
    await fs.rm(TEMP_DIR, { recursive: true, force: true });
    console.log('Temporary files cleaned up');
  } catch (error) {
    console.error('Failed to clean temp files:', error);
  }
};

/**
 * 下载pdf，oss上传，mysql插入，发起ragflow schedule job
 */
export const handleWeb2Pdf = async (request, page, siteCode, siteName) => {
  logger.info(`-----${siteCode}: start to handleWebPage ${request.url}-----`)
  // 获取网页基本信息
  const siteDomain = extractDomain(request.url);
  const pageUrl = request.url;
  // const pageContent = await page.content();
  
  if (pageUrl.indexOf('.pdf') > -1  // nifdc特殊页面, 仅保存页面的pdf
  && (pageUrl.indexOf('search.do?formAction=openUrl') > -1 || pageUrl.indexOf('pqf/file_path/help/') > -1)) {
    
    const existingRecord = await WebPage.getEntityByPageUrl(pageUrl);
    if (!!existingRecord && existingRecord != null) {
      const downloadDir = path.resolve(ATTACHMENT_DIR);
      logger.debug(`${JSON.stringify(existingRecord)}`);
      fs.mkdirSync(downloadDir, { recursive: true });
      const attachmentInfo = await saveAttachementFile(pageUrl, existingRecord.page_title + ".pdf", downloadDir, null, siteDomain, siteName, siteCode);
    logger.info(`finish attach download1-`);

      let extra = existingRecord.extra;
      extra.fileKey = attachmentInfo.ossKey;
      extra.filename = attachmentInfo.fileName;

      const webPage = {
        id: existingRecord.id,
        pageContent: null,
        pageType: 'FILE',
        extra: JSON.stringify(extra),
        status: 'DONE'
      };
      await WebPage.updatePage(webPage)
      logger.info(`${siteCode} WebPage updated, pageId=${existingRecord.id}`);

      // API发起ragflow写入任务
      // if (ids.length > 0) {
        // await createYiyaJob([existingRecord.id]);
      // }
    }
  } else { // 常规叶子页面的爬取，并保存
    // logger.info(`>>>>${siteCode}--1: im visiting page now! title=${pageTitle}, url=${pageUrl}`);
    await page.waitForSelector('h2');
    // logger.info(`>>>>${siteCode}: im visiting page now! title=${pageTitle}, url=${pageUrl}`);
    
    const pageTitle = await page.title();
    const pageType = SPIDER_SITE_TYPE.page;
    const pdfName = sanitizeFilename(pageTitle) + ".pdf";
    const localPdfPath = path.join(TEMP_DIR, pdfName);
    const ossKey = `${process.env.OSS_SUBDIR || 'spider'}/${pdfName}`;
      
    // 检查该页面是否已经导入数据库，是、则退出
    let pageId;
    let needUpdate = true;
    const existingRecord = await WebPage.getEntityByPageUrl(pageUrl);
    logger.info(`${siteCode}: existingRecord=${existingRecord.id}`);
    if (!!existingRecord || null != existingRecord) {
      pageId = existingRecord.id;
      if (existingRecord.status == 'NEW') {
        logger.info(`${siteCode}: this is a new page-00000, ${pageUrl}`);
      } else {
        logger.info(`${siteCode}: this page has alreay been imported before, pageId=${existingRecord.id}`);
        return;
      }
    } else {
      needUpdate = false;
      logger.info(`${siteCode}: this is a new page, ${pageUrl}`)
    }

    logger.debug("start to save page into pdf");
    
    // 开始下载流程
    try {
      await convertPageToLocalPdf(page, localPdfPath);
      logger.info(`${siteCode} WebPage PDF saved locally: ${localPdfPath}`);
      
      // 上传到OSS
      await uploadToOSS(localPdfPath, ossKey);

      if (needUpdate) {
        let extra = existingRecord.extra;
        extra.fileKey = ossKey;
        extra.filename = pdfName;

        const pageMetaNameEle = await page.$('meta[name="ColumnName"]');
        if (pageMetaNameEle) {
          extra.pageMetaName = page.$eval('meta[name="ColumnName"]', (el) => el.content);
        }
        const pageMetaDescriptionEle = await page.$('meta[name="ColumnDescription"]');
        if (pageMetaDescriptionEle) {
          extra.pageMetaDescription = page.$eval('meta[name="ColumnDescription"]', (el) => el.content);
        }

        logger.info(`updatePage: pageId: ${pageId}, extra: ${extra}`);

        const webPage = {
          id: pageId,
          pageContent: null,
          extra: JSON.stringify(extra),
          pageType: 'PAGE',
          status: 'DONE'
        };
        await WebPage.updatePage(webPage)
        logger.info(`${siteCode} WebPage updated, pageId=${pageId}`);
      } else { //insert
        let extra = {
          fileKey: ossKey,
          filename: pdfName
        };

        const pageMetaNameEle = await page.$('meta[name="ColumnName"]');
        if (pageMetaNameEle) {
          extra.pageMetaName = page.$eval('meta[name="ColumnName"]', (el) => el.content);
        }
        const pageMetaDescriptionEle = await page.$('meta[name="ColumnDescription"]');
        if (pageMetaDescriptionEle) {
          extra.pageMetaDescription = page.$eval('meta[name="ColumnDescription"]', (el) => el.content);
        }
        // 保存记录到数据库
        const webPage = {
          siteDomain: siteDomain, 
          siteName: siteName, 
          siteCode: siteCode, 
          pageUrl: pageUrl, 
          pageType: pageType, 
          pageTitle: pageTitle, 
          pageContent: null, 
          extra: JSON.stringify(extra),
          parentId: null,
          status: 'DONE'
        };
        pageId = await WebPage.save(webPage);
        logger.info(`${siteCode} WebPage saved to mysql, pageId=${pageId}`);
      }
      
      let ids = await attachmentCrawler(request, page, pageId, siteDomain, siteName, siteCode);
      // ids = attachmentIds;
      ids.concat(pageId);

      // API发起ragflow写入任务
      // if (ids.length > 0) {
      //   await createYiyaJob(ids);
      // }
  
    } catch (error) {
      logger.error(`Failed to process ${request.url}:`, error);
    }
  }
}

/**
 * 处理文件名中包含特殊字符，例如斜杠
 * @param {*} filename 
 * @returns 
 */
function sanitizeFilename(filename) {
  return filename.replace(/[\\/:*?"<>|.]/g, '_');
}

export const createNewPage = async (siteDomain, siteName, siteCode, pageUrl, dataItem) => {
  // 检查该页面是否已经导入数据库，是、则退出
  const existingPage = await WebPage.getEntityByPageUrl(pageUrl);
  // logger.debug(JSON.stringify(existingPage))
  if (!!existingPage || null != existingPage) {
    if (existingPage.status != 'NEW') {
      logger.info(`${siteCode} - pageUrl=${pageUrl}: it is an existing page, id=${existingPage.id}`);
      return;
    } else {
      logger.info(`${siteCode} - waiting for crawlling: ${pageUrl}`);
    }
  } else {
    logger.info(`createNewPage of ${siteCode}: pageUrl=${pageUrl}`);
    if (existingPage == null) {
      // 保存记录到数据库
      const webPage = {
        siteDomain: siteDomain, 
        siteName: siteName,//SITE_CONFIG.nmpa.name, 
        siteCode: siteCode,//SITE_CONFIG.nmpa.code, 
        pageUrl: pageUrl, 
        pageType: SPIDER_SITE_TYPE.page, 
        pageTitle: dataItem.title, 
        pageContent: null, 
        extra: JSON.stringify({ releaseDate: dataItem.date, href: dataItem.href, title: dataItem.title, parentUrl: dataItem.parentUrl }),
        parentId: null,
        status: 'NEW'
      };
      const pageId = await WebPage.save(webPage);
      logger.info(`done createNewPage of ${siteCode}: pageUrl=${pageUrl}, pageUrl=${pageUrl}, pageId=${pageId}`);
    } else {
      logger.info(`${siteCode} - ${pageUrl} is an exsiting spider page record, but not yet crawled completely`);
    }
  }
}

// 解析页面上的附件
const attachmentCrawler = async (request, page, parentId, siteDomain, siteName, siteCode) => {
  let ids = [];
  const docLinks = await page.$$eval('a', (anchors, baseUrl) => {
    return anchors
      .filter(a => a.href.match(/\.(docx?|pdf|xlsx|xls|doc|ppt)$/i))
      .map(a => ({
        href: new URL(a.href, baseUrl).href,
        text: a.textContent.trim()
      }));
  }, request.loadedUrl);
  
  logger.info(`${siteCode}: parentId(${parentId}) has docLinks: ${docLinks}`);

  if (docLinks.length > 0) {
    logger.info(`${siteCode}: start to download attachments`);
    // 确保下载目录存在
    const downloadDir = path.resolve(ATTACHMENT_DIR);
    fs.mkdirSync(downloadDir, { recursive: true });

    for (const { href, text } of docLinks) {
      try {
        const attachmentInfo = await saveAttachementFile(href, text, downloadDir, request.loadedUrl, parentId, siteDomain, siteName, siteCode);
        logger.info(`finish attach download2-`);
        // 保存附件到mysql
        const attachmentEntity = {
          parentId: parentId,
          siteDomain: siteDomain,
          siteName: siteName,
          siteCode: siteCode,
          pageUrl: href,
          pageType: SPIDER_SITE_TYPE.file,
          pageTitle: attachmentInfo.fileName,
          pageContent: null,
          extra: JSON.stringify({
            fileKey: attachmentInfo.ossKey,
            filename: attachmentInfo.fileName
          }),
          status: 'DONE'
        };
        logger.info(`New attachment: ${attachmentEntity} of parentId(${parentId})`);
        const fileId = await WebPage.save(attachmentEntity);

        logger.info(`Attachment(parentId=${parentId}, attachmentId=${fileId}) saved to mysql`);
        // return fileId;
        ids.push(fileId);
      } catch (error) {
        logger.error(`下载失败：${href}，错误信息：${error.message}`);
      }
    }
  }
  return ids;
}

const saveAttachementFile = async (href, text, downloadDir, reqUrl, parentId, siteDomain, siteName, siteCode) => {
  const url = reqUrl != null ? new URL(href, reqUrl).toString() : href;
  const fileName = `${sanitizeFilename(text).split(".")[0]}${path.extname(url)}`;
  const filePath = path.join(downloadDir, fileName);

  logger.info(`正在下载：${url}, filePath:${filePath}`);

  const response = await got.stream(url);
  await streamPipeline(response, fs.createWriteStream(filePath));
  logger.info(`已保存：${filePath}`);

  const ossKey = `spider/attachment/${fileName}`;
  // 上传附件到oss
  await uploadToOSS(filePath, ossKey);

return {fileName, filePath, ossKey}
  
}

const convertPageToLocalPdf = async (page, localPdfPath) => {
  // 调整页面设置
    await page.setViewport({ width: 1920, height: 1080 });
    
    await page.addStyleTag({
        content: `
            body {
                overflow: visible !important;
                height: auto !important;
            }
            html, body {
                width: 100% !important;
            }
            @media print {
                body {
                    -webkit-print-color-adjust: exact;
                    print-color-adjust: exact;
                }
            }
        `,
    });

    // 生成PDF
    await page.pdf({
      path: localPdfPath,
      format: 'A2',
      printBackground: true,
      fullPage: true,
      margin: {
        top: '5mm',
        right: '5mm',
        bottom: '5mm',
        left: '5mm'
      }
    });
    
    
}

File: nmpa_crawler/utils/fileHandler.py
# -*- coding: utf-8 -*-
"""
文件处理器 - 完全仿照 yiya-crawler 的 fileHandler.js
"""

import os
import json
import time
import asyncio
from typing import Dict, Any, Optional
from urllib.parse import urlparse
from datetime import datetime

from utils.logger import logger
from config.constants import SPIDER_SITE_TYPE


# 临时PDF存储目录 - 完全复制自 yiya-crawler
TEMP_DIR = 'downloads'
ATTACHMENT_DIR = 'downloads/attachments'


def ensure_dir(dir_path: str):
    """确保目录存在 - 复制自 yiya-crawler"""
    try:
        os.makedirs(dir_path, exist_ok=True)
    except Exception as e:
        logger.error(f"创建目录失败 {dir_path}: {e}")


def generate_unique_name(url: str, ext: str = 'pdf') -> str:
    """生成唯一的文件名 - 复制自 yiya-crawler"""
    timestamp = int(time.time() * 1000)
    url_hash = abs(hash(url)) % 1000000
    return f"{timestamp}_{url_hash}.{ext}"


def generate_pdf_name(url: str) -> str:
    """生成唯一文件名 - 复制自 yiya-crawler"""
    timestamp = int(time.time() * 1000)
    hostname = urlparse(url).hostname
    return f"{hostname}_{timestamp}.pdf"


def sanitize_filename(filename: str) -> str:
    """处理文件名中的特殊字符 - 复制自 yiya-crawler"""
    import re
    return re.sub(r'[\\/:*?"<>|]', '_', filename)


async def cleanup_temp_files():
    """清理临时文件 - 复制自 yiya-crawler"""
    try:
        import shutil
        if os.path.exists(TEMP_DIR):
            shutil.rmtree(TEMP_DIR, ignore_errors=True)
            logger.info("Temporary files cleaned up")
    except Exception as error:
        logger.error(f"Failed to clean temp files: {error}")


async def handleWeb2Pdf(request, page, site_code: str, site_name: str):
    """处理网页到PDF的转换 - 完全仿照 yiya-crawler 的 handleWeb2Pdf"""
    logger.info(f"-----{site_code}: start to handleWebPage {request.url}-----")

    # 获取网页基本信息 - 复制自 yiya-crawler
    try:
        from urllib.parse import urlparse
        site_domain = urlparse(request.url).netloc
        page_url = request.url
    except:
        site_domain = "www.nmpa.gov.cn"
        page_url = request.url

    # 检查是否是PDF文件 - 复制自 yiya-crawler 的逻辑
    if page_url.endswith('.pdf') and ('search.do?formAction=openUrl' in page_url or 'pqf/file_path/help/' in page_url):
        logger.info(f"{site_code}: 处理PDF文件: {page_url}")
        # 这里可以添加PDF下载逻辑，暂时跳过
        return

    # 常规叶子页面的爬取，并保存 - 完全仿照 yiya-crawler
    try:
        # 等待页面加载
        try:
            await page.wait_for_selector('h2', timeout=10000)
        except:
            logger.warning(f"{site_code}: 等待h2元素超时")

        page_title = await page.title()
        page_type = SPIDER_SITE_TYPE["page"]
        pdf_name = sanitize_filename(page_title) + ".pdf"
        local_pdf_path = os.path.join(TEMP_DIR, pdf_name)

        logger.info(f"{site_code}: 页面标题: {page_title}")
        logger.info(f"{site_code}: PDF文件名: {pdf_name}")

        # 检查页面是否已存在 - 简化版本
        existing_file = os.path.join(TEMP_DIR, f"{page_title}.json")
        need_update = True

        if os.path.exists(existing_file):
            logger.info(f"{site_code}: 页面已存在，跳过: {page_url}")
            return

        logger.debug(f"{site_code}: 开始保存页面数据")

        # 开始数据提取流程 - 仿照 yiya-crawler 的逻辑
        try:
            # 提取页面数据
            page_data = await extract_page_data(page, page_url, site_code, site_name, page_title)

            # 保存页面数据
            await save_page_data(page_data, page_url, site_code)

            logger.info(f"{site_code}: 页面数据保存成功: {page_url}")

        except Exception as error:
            logger.error(f"{site_code}: 处理页面失败 {page_url}: {error}")

    except Exception as e:
        logger.error(f"{site_code}: 处理网页失败: {e}")


async def extract_page_data(page, page_url: str, site_code: str, site_name: str, page_title: str) -> Dict[str, Any]:
    """提取页面数据 - 仿照 yiya-crawler 的数据提取逻辑"""

    # 提取页面文本内容
    try:
        page_text = await page.evaluate('() => document.body.innerText')
    except:
        page_text = ""

    # 提取页面HTML内容
    try:
        page_html = await page.content()
    except:
        page_html = ""

    # 提取元数据 - 仿照 yiya-crawler 的 meta 提取
    meta_data = {}

    try:
        # 提取 meta 标签
        meta_elements = await page.query_selector_all('meta')
        for element in meta_elements:
            name = await element.get_attribute('name')
            content = await element.get_attribute('content')
            if name and content:
                meta_data[name] = content
    except:
        pass

    # 提取发布时间
    publish_time = ""
    time_selectors = [
        'meta[name="date"]',
        'meta[name="publish_date"]',
        'meta[property="article:published_time"]',
        '.time',
        '.date',
        '.publish-time'
    ]

    for selector in time_selectors:
        try:
            if selector.startswith('meta'):
                element = await page.query_selector(selector)
                if element:
                    publish_time = await element.get_attribute('content')
                    break
            else:
                element = await page.query_selector(selector)
                if element:
                    publish_time = await element.inner_text()
                    break
        except:
            continue

    # 构建页面数据 - 仿照 yiya-crawler 的数据结构
    page_data = {
        "site_domain": urlparse(page_url).netloc,
        "site_name": site_name,
        "site_code": site_code,
        "page_url": page_url,
        "page_type": SPIDER_SITE_TYPE["page"],
        "page_title": page_title,
        "page_content": page_text[:1000] + "..." if len(page_text) > 1000 else page_text,  # 限制长度
        "page_html": page_html[:2000] + "..." if len(page_html) > 2000 else page_html,
        "extra": json.dumps({
            "publish_time": publish_time,
            "meta_data": meta_data,
            "crawl_time": datetime.now().isoformat()
        }),
        "parent_id": None,
        "status": "DONE",
        "created_time": datetime.now().isoformat()
    }

    return page_data


async def save_page_data(page_data: Dict[str, Any], page_url: str, site_code: str):
    """保存页面数据 - 仿照 yiya-crawler 的保存逻辑"""

    # 确保目录存在
    ensure_dir(TEMP_DIR)

    # 生成文件名
    filename = generate_unique_name(page_url, 'json')
    filepath = os.path.join(TEMP_DIR, filename)

    # 保存数据
    try:
        with open(filepath, 'w', encoding='utf-8') as f:
            json.dump(page_data, f, ensure_ascii=False, indent=2, default=str)

        logger.info(f"{site_code}: 页面数据已保存: {filepath}")

    except Exception as e:
        logger.error(f"{site_code}: 保存页面数据失败: {e}")


async def createNewPage(site_domain: str, site_name: str, site_code: str, pageUrl: str, dataItem: Dict[str, Any]):
    """创建新页面记录 - 完全仿照 yiya-crawler 的 createNewPage"""

    # 检查该页面是否已经导入 - 简化版本
    existing_file = os.path.join(TEMP_DIR, f"{dataItem.get('title', 'unknown')}.json")

    if os.path.exists(existing_file):
        logger.info(f"{site_code} - pageUrl={pageUrl}: 页面已存在，跳过")
        return

    logger.info(f"createNewPage of {site_code}: pageUrl={pageUrl}")

    # 保存记录到文件 - 仿照 yiya-crawler 的数据结构
    web_page_data = {
        "site_domain": site_domain,
        "site_name": site_name,
        "site_code": site_code,
        "page_url": pageUrl,
        "page_type": SPIDER_SITE_TYPE["page"],
        "page_title": dataItem.get("title", ""),
        "page_content": None,
        "extra": json.dumps({
            "releaseDate": dataItem.get("date", ""),
            "href": dataItem.get("href", ""),
            "title": dataItem.get("title", ""),
            "parentUrl": dataItem.get("parentUrl", "")
        }),
        "parent_id": None,
        "status": "NEW",
        "created_time": datetime.now().isoformat()
    }

    # 保存数据
    filename = generate_unique_name(pageUrl, 'json')
    filepath = os.path.join(TEMP_DIR, filename)

    try:
        with open(filepath, 'w', encoding='utf-8') as f:
            json.dump(web_page_data, f, ensure_ascii=False, indent=2, default=str)

        logger.info(f"{site_code}: 新页面记录已保存: {filepath}")

    except Exception as e:
        logger.error(f"{site_code}: 保存新页面记录失败: {e}")